### PR TITLE
Call updates and multi-use consumables

### LINKED ISSUE #1: Feedback on Mod.Call
There is alot of room for improvement regarding the Mod.Call API.

## "AddMod"
Is unnecessary.
The tab can be automatically generated from the first call of a new item being added via "AddConsumable". If the mod hasn't been detected yet, generate the tab, else, do nothing.
The mod icon itself can be determined from the mod the item belongs to, you can see it here: <https://github.com/tModLoader/tModLoader/blob/1.4.4/patches/tModLoader/Terraria/GameContent/Bestiary/ModSourceBestiaryInfoElement.cs#L20>. As it's optional, a backup texture is required, I think it's just a "?" in tml.

## "AddConsumable"
I propose to replace the current API with support for versioning and cleaner parameters.
The issues of the current the current API are as follows:
* Not localizable
* Not flexible (modder has to change tooltips, names, etc. in multiple locations)
* Icon dimensions are hardcoded, modder has to be aware to update them whenever the texture changes

### How to extend this API to remedy the above points?
First, after "AddConsumable", the next parameter should be of type `Mod`: This will allow you to direct any logging or errors to the mod directly, and lets you populate the tab using `Mod.DisplayName`, see "AddMod" above.
Then, the next arg should be of type `string` representing the mod version of **this** mod. This allows you to
1. differentiate the new API from old API (if current version => new/current API, else old API(s))
2. make sure the use of old versions of the API are detected and can be acted upon (i.e. reject, or limited support)

You can see how it's cast to `System.Version` here for convenience: <https://github.com/SamsonAllen13/ClickerClass/blob/master/ClickerModCalls.cs#L26> (even though it's unused, but usage would be like: `version > new Version(0, 1, 2)`)

Localization is done by handling every `string` as a `LocalizedText` instead. This lets the game update all the texts dynamically whenever the language is changed. As a bonus, it supports custom formatting, and reusability (modder just has to change the text in their localization file, and it will update in Munchies automatically). More on this later and how the current codebase has to be changed to support it.

Icon dimensions can be pulled from the item itself, and done so on demand as Asset<Texture2D>, else it would require loading the texture on the Mod.Call which stalls loading considerably.

Now onto the parameters. Almost all of them can be inferred from the `ModItem` instance of a given item automatically. Example using current API, and a wrapper method over the call:
```cs
AddItem(ModContent.GetInstance<MyItem>(), "player_normal", () => Main.LocalPlayer.statLifeMax >= 500);

public void AddItem(ModItem modItem, string mode, Func<bool> consumed)
{
	var texture = ModContent.Request<Texture2D>(modItem.Texture, ReLogic.Content.AssetRequestMode.ImmediateLoad);
	(string width, string height) = (texture.Width().ToString(), texture.Height().ToString());
	munchiesMod.Call("AddConsumable",
		modItem.Mod.DisplayName,
		modItem.DisplayName.ToString(),
		modItem.Texture,
		modItem.Tooltip.ToString(),
		mode,
		consumed,
		width,
		height);
}
```

As you can see, it has been fully reduced to just `ModItem`,`string`, and `Func<bool>`. Now clearly, this is just a workaround, and some of the things there are bad practice, like loading the texture using `ImmediateLoad` to get the dimensions, or doing `.ToString()` for the current language of the texts. The mod name would have to be replaced with the `Mod` and Version `string` too.

This means all you need is those three parameters for a minimal call. If the modder wishes to modify the parameters, you can have optional overrides after the three parameters, which will take precedence on registering the item (i.e. texture is 20x20 but modder for some reason wants it to be 32x32, this lets the modder set it). You can check for the optional parameters by checking the length of the `args` and evaluating any that aren't `null` (the modder fills in any optional ones that are skipped with `null`).

### Backend
There is alot of hardcoding the vanilla IDs and texts. You can access the item name text with `Lang.GetItemName(id)`, the tooltip with `Language.GetText($"ItemTooltip.{ItemID.Search.GetName(id)}")`, textures with `$"Terraria/Images/Item_{id}"`.
Since the textures are used for UI, and the dimensions are important, to prevent increasing load times, you can generate the UI on opening it for the first time ingame. I did not do a deep dive into the code so I can't say anything else.

tl;dr: Textures should be stored as `Asset<Texture2D>` (and dimensions inferred from that, or the override), texts as `LocalizedText`. 

To support mods adding vanilla items to this (if they rework them for example), you can have an "AddVanillaConsumable" call that takes in an `int` id instead of `ModItem`, and does some calculations mentioned above, plus additional overrides.

## Backwards compatibility
This change will break existing mods relying on the flawed API. It's up to you to decide to continue to support it or not. As the mod is fairly new, I think it's justified to break the API early and then continue with one that is more flexible and reliable.

## Proposed "AddConsumable" Mod.Call
```cs
munchiesMod.Call("AddConsumable", Mod/this, "1.2", ModContent.GetInstance<MyModItem>(), "player_normal", myFuncboolCondition);
```

### PR DESCRIPTION
**Pull request by jbossjaslow**
Overhauled the Call function to better handle mod items and streamline the process for developers. Also added multi-use consumables as a config option

### CODE REVIEW

File: Models/Consumable.cs [hunk 1]
```diff
@@ -1,79 +1,80 @@
...
-			SetTextureAndDimensions(texturePath: vc.TexturePath, assetDimensions: vc.AssetDimensions);
-		}
+		#region Public helpers
+		public bool HasBeenConsumed => CurrentCount() == TotalCount();
+		public bool IsMultiUse => TotalCount() > 1;
+		#endregion
 
-		private void SetTextureAndDimensions(string texturePath, (float X, float Y) assetDimensions) {
+		#region Private helpers
+		private void SetTextureAndDimensions(string texturePath) {
```
Comment (direwolf420) [inline Models/Consumable.cs]: Can be renamed to just `SetTexture`. Same with the other method of the same name

File: Models/Consumable.cs [hunk 2]
```diff
@@ -1,79 +1,80 @@
...
-			//TexturePath = texturePath;
-			//AssetDimensions = assetDimensions;
-			HoverText = hoverText;
-			Type = type;
-			CustomTextColor = null;
-			HasBeenConsumed = hasBeenConsumed;
+		#region Init
+		public Consumable(ModItem modItem, object CategoryOrCustomColor, Func<int> currentCount, Func<int> totalCount) {
+			DisplayText = modItem.DisplayName.ToString();
+			HoverText = modItem.Tooltip.ToString();
```
Comment (direwolf420) [inline Models/Consumable.cs]: This is not solving the localization problem outlined in the issue. Invoking .Tostring() or .Value fetches the string for the currently selected language. Changing language then won't change these until next mod reload. To remedy this, don't call these methods, and instead store DisplayText/HoverText as a `LocalizedText`

File: Models/Consumable.cs [hunk 3]
```diff
@@ -1,79 +1,80 @@
...
-			//AssetDimensions = assetDimensions;
-			HoverText = hoverText;
-			Type = null;
-			CustomTextColor = customHoverColor;
-			HasBeenConsumed = hasBeenConsumed;
-
-			SetTextureAndDimensions(texturePath: texturePath, assetDimensions: assetDimensions);
+		public Consumable(int vanillaItemId, ConsumableType type, Func<int> currentCount, Func<int> totalCount) {
+			DisplayText = Lang.GetItemName(vanillaItemId).Value;
+			HoverText = Language.GetText($"ItemTooltip.{ItemID.Search.GetName(vanillaItemId)}").Value;
```
Comment (direwolf420) [inline Models/Consumable.cs]: Same as above

File: Models/ConsumableMod.cs [hunk 4]
```diff
@@ -12,10 +14,32 @@ public class ConsumableMod {
...
+			SetTextureAndDimensionsForVanilla(modTabTexturePath);
 		}
 
-		private void SetTextureAndDimensions(string texturePath) {
+		public ConsumableMod(Mod mod) {
+			ModTabName = mod.DisplayNameClean;
+			SetTextureAndDimensionsForExternalMod(mod);
+		}
+
+		private void SetTextureAndDimensionsForExternalMod(Mod mod) {
```
Comment (direwolf420) [inline Models/ConsumableMod.cs]: `SetTextureForExternalMod`

File: Models/ConsumableMod.cs [hunk 5]
```diff
@@ -12,10 +14,32 @@ public class ConsumableMod {
...
+			if (mod.HasAsset("icon_small")) {
+				Asset<Texture2D> texture = mod.Assets.Request<Texture2D>("icon_small");
+				if (texture.Size() == new Vector2(30)) {
+					ModTabTexture = texture;
+					UsingMissingTexture = false;
+				} else {
+					mod.Logger.Info("icon_small needs to be 30x30 pixels");
```
Comment (direwolf420) [inline Models/ConsumableMod.cs]: The logging is a relic of tml copypaste, should be removed. It should be mentioned in the documentation for the calls that icon_small is used for the tab. It will encourage modders to add it themselves

File: Models/Report.cs [hunk 6]
```diff
@@ -10,62 +10,37 @@ public class Report {
...
-					}
+		public static bool AddConsumableToList(ConsumableMod mod, Consumable consumable) {
+			ConsumablesEntry entry = GetModEntryOrAddIfNeeded(mod);
 
-					// consumable does not exist, add it
-					entry.Consumables.Add(consumable);
-					return true;
+			foreach (Consumable c in entry.Consumables) {
+				if (c.DisplayText == consumable.DisplayText) {
+					// consumable already exists, exit
```
Comment (direwolf420) [inline Models/Report.cs]: Probably better to store a unique identifier (such as the item's type (ModItem.Type / id) to then check for equality. This may also cause collisions if two mods add an item with the same name

File: Models/VanillaConsumable.cs [hunk 7]
```diff
@@ -10,6 +10,11 @@ public VanillaConsumable(ConsumableItem item) {
 		}
 
 		public ConsumableType Type => Item switch {
```
Comment (direwolf420) [inline Models/VanillaConsumable.cs]: This whole class has alot of hardcode which I don't think is necessary. You can compress the DisplayText to just `Lang.GetItemName(vanillaItemId)`.
Not sure about the HoverText since you add some extra stuff to it. In general, any strings you are using that should be displayed to users are supposed to go into the localization file (like the config does at the moment). You can read more about how to add localization to your mod here: <https://github.com/tModLoader/tModLoader/wiki/Localization>.
ConsumableItem -> ItemId is backwards design imo, it should be sufficient to just have them referenced as the ItemId directly as it's vanilla, and use the consts (`ItemID.LifeCrystal` instead of 29 etc.)
My impression is that all of these could be replaced by using the `Consumable` class which you already have. You can use the API in your own mod aswell!

File: Munchies.cs [hunk 8]
```diff
@@ -29,112 +29,85 @@ public override void Unload() {
...
+					CategoryOrCustomColor: args[4],
+					currentCount: currentCount,
+					totalCount: totalCount
+				);
+
+				return Report.AddConsumableToList(mod: externalMod, consumable: consumable);
+
+				// Add optional fields for max # of items, option of func<int> for how many have been consumed
+			} catch (Exception e) {
+				(mod ?? this).Logger.Error($"Error adding consumable: {e.Message}, {e.StackTrace}");
```
Comment (direwolf420) [inline Munchies.cs]: Two things:
1. The logger should still originate from your own mod always (`this`), since you invoke the logging, and the mod's Name should be used in the log message to indicate/point to it
2. Logging the exception's data is redundant as TML hooks into the exception logging which bypasses catching, so it always prints these out

File: Munchies.cs [hunk 9]
```diff
@@ -29,112 +29,85 @@ public override void Unload() {
...
+				consumable = new(
+					vanillaItemId: itemId,
+					type: GetType(args[4] as string),
+					currentCount: currentCount,
+					totalCount: totalCount
+				);
+
+				return Report.AddConsumableToList(mod: vanillaMod, consumable: consumable);
+			} catch (Exception e) {
+				Logger.Error($"Error adding consumable: {e.Message}, {e.StackTrace}");
```
Comment (direwolf420) [inline Munchies.cs]: Same as above

**Review by direwolf420 — COMMENTED** [review]
As an aside, whenever you do `Request<Texture2D>` to use in UIs (for fetching dimensions primarily), you should use it with `ReLogic.Content.AssetRequestMode.ImmediateLoad`. Otherwise it can happen that the texture is not loaded when accessing dimensions and it returns 0, causing issues in the UI
